### PR TITLE
Test porting

### PR DESCRIPTION
**Pull request by larryk85**
#DAWN-390,391,392

 Porting api tests, memory tests, and extended memory tests

### CODE REVIEW

File: contracts/eosiolib/crypto.h [hunk 1]
```diff
@@ -14,10 +14,76 @@ extern "C" {
...
+/**
+ *  This method is implemented as:
+ *
+ *  checksum calc_hash;
+ *  ripemd160( data, length, &calc_hash );
+ *  eos_assert( calc_hash == hash, "invalid hash" );
+ *
+ *  This method is optimized to a NO-OP when in fast evaluation mode
+ */
+void eos_assert_ripemd160( char* data, uint32_t length, const checksum_base* hash );
```
Comment (heifner) [inline contracts/eosiolib/crypto.h]: These need to be all eosio_*

File: tests/wasm_tests/wasm_tests.cpp [hunk 2]
```diff
@@ -264,6 +267,163 @@ BOOST_FIXTURE_TEST_CASE( test_api_bootstrap, tester ) try {
       BOOST_CHECK_EQUAL(transaction_receipt::executed, receipt.status);
    }
 } FC_LOG_AND_RETHROW() /// test_api_bootstrap
+#endif
+
+BOOST_FIXTURE_TEST_CASE( test_proxy, tester ) try {
```
Comment (elmato) [inline tests/wasm_tests/wasm_tests.cpp]: proxy contract tests are now in currency_test.cpp

File: libraries/chain/include/eosio/chain/apply_context.hpp [hunk 3]
```diff
@@ -774,10 +774,9 @@ using apply_handler = std::function<void(apply_context&)>;
 
       if( itr == idx.end() ||
           itr->t_id != t_id.id ||
-          !impl::record_scope_compare<IndexType, Scope>::compare(*itr, keys)) return -1;
+          !impl::record_scope_compare<IndexType, Scope>::compare(*itr, keys)) return 0;
```
Comment (elmato) [inline libraries/chain/include/eosio/chain/apply_context.hpp]: shouldn't we keep the -1 return value? 
How we can distinguish between reading a zero sized record (besides keys) and a "record not found" error.

File: libraries/chain/include/eosio/chain/apply_context.hpp [hunk 4]
```diff
@@ -799,7 +798,7 @@ using apply_handler = std::function<void(apply_context&)>;
 
       auto itr = idx.lower_bound( tuple );
       if( itr == idx.end() ||
-         itr->t_id != t_id.id ) return -1;
+         itr->t_id != t_id.id ) return 0;
```
Comment (elmato) [inline libraries/chain/include/eosio/chain/apply_context.hpp]: Same as above

File: contracts/eosiolib/checktime.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+extern "C" {
+  /**
+   *  @defgroup checktime API
+   *  @brief Declares checktime function. 
+   *  @ingroup checktimeapi
+   *
+   *  @{
+   */
+   void checktime();
```
Comment (wanderingbort) [inline contracts/eosiolib/checktime.h]: I think its preferable to not treat `chectime` as part of the ABI.  I don't think we want contracts calling this directly. @spoonincode may want to weigh in.

File: contracts/test_api/test_account.cpp [hunk 6]
```diff
@@ -4,29 +4,43 @@
...
 
-unsigned int test_account::test_balance_acc1()
-{
+void test_account::test_balance_acc1() {
    eosio::print("test_balance_acc1\n");
-   eosio::account::account_balance b;
-
+   //eosio::account::account_balance b;
+   
+/*
```
Comment (wanderingbort) [inline contracts/test_api/test_account.cpp]: Go ahead and delete this test as opposed to merely commenting it out.

File: contracts/test_api/test_privileged.cpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+#include "test_api.hpp"
+
+/*
+void test_privileged::set_privileged() {
+   prints("hello\n");
+}
+*/
+
+void test_privileged::test_is_privileged() {
+   prints("HELLO\n");
```
Comment (wanderingbort) [inline contracts/test_api/test_privileged.cpp]: we should probably have these tests run silently instead of printing out debugging totems.

File: contracts/test_api/test_transaction.cpp [hunk 8]
```diff
@@ -4,106 +4,206 @@
...
-   for (int i = 0; i < 10; i++) {
-      transaction_create();
-   }
+void test_transaction::test_transaction_size() {
+   uint32_t trans_size = 0;
+   read_action( (char*)&trans_size, sizeof(uint32_t) );
+   
+   eosio_assert( trans_size == transaction_size(), "transaction size does not match" );
+}
+/*
```
Comment (wanderingbort) [inline contracts/test_api/test_transaction.cpp]: this looks like a stale copy of the test? can it be removed?

File: contracts/test_api_db/test_api_db.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+      WASM_TEST_HANDLER(test_db, primary_i64_general);
+      WASM_TEST_HANDLER(test_db, primary_i64_lowerbound);
+      WASM_TEST_HANDLER(test_db, primary_i64_upperbound);
+      WASM_TEST_HANDLER(test_db, idx64_general);
+      WASM_TEST_HANDLER(test_db, idx64_lowerbound);
+      WASM_TEST_HANDLER(test_db, idx64_upperbound);
+
+      WASM_TEST_HANDLER(test_db, key_str_general);
+#if 0
```
Comment (wanderingbort) [inline contracts/test_api_db/test_api_db.cpp]: since we are deprecating the old API, we can go ahead and remove these stubs too

File: contracts/test_api_db/test_db.cpp [hunk 10]
```diff
@@ -64,114 +64,122 @@ extern "C" {
...
+     for (uint32_t i = 0; i < n; i++) {
+        if (c1[i] != c2[i]) {
+           return false;
+        }
+     }
+     return true;
+  }
 }
 
 #if 0
```
Comment (wanderingbort) [inline contracts/test_api_db/test_db.cpp]: remove stale/old tests

File: contracts/test_api_mem/test_memory.cpp [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+void verify_mem(const void* const ptr, const uint32_t val, const uint32_t size)
+{
+   const char* char_ptr = (const char*)ptr;
+   for (uint32_t i = 0; i < size; ++i)
+   {
+      eosio_assert(static_cast<uint32_t>(static_cast<unsigned char>(char_ptr[i])) == val, "buf slot doesn't match");
+   }
+}
+
+/*
```
Comment (wanderingbort) [inline contracts/test_api_mem/test_memory.cpp]: this looks like a debugging helper function, can it be deleted?

File: libraries/chain/wasm_interface.cpp [hunk 12]
```diff
@@ -43,27 +45,27 @@ using boost::asio::io_service;
       /**
       * Name of the account who's balance this is
       */
-      account_name account;
+      eosio::chain::account_name account;
```
Comment (wanderingbort) [inline libraries/chain/wasm_interface.cpp]: go ahead and remove this if 0'd section

File: libraries/test.err [hunk 13]
```diff
@@ -0,0 +1 @@
+/Users/judgefudge/scripts/run_tests: line 7: /Users/judgefudge/bucky_eos/eos/libraries/chain_test: No such file or directory
```
Comment (wanderingbort) [inline libraries/test.err]: gonna guess this wasn't supposed to get checked in ... judgefudge :smirk:

Comment (wanderingbort) [reply libraries/test.err]: nor the test.out which is empty

File: tests/api_tests/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,15 @@
+#add_wast_target(test_api "${CMAKE_SOURCE_DIR}/contracts" ${CMAKE_CURRENT_BINARY_DIR})
```
Comment (wanderingbort) [inline tests/api_tests/CMakeLists.txt]: is this file necessary if they are all commented out?

File: tests/common/database_fixture.hpp [hunk 15]
```diff
@@ -4,12 +4,14 @@
  */
 #pragma once
 
-#include <eos/chain/chain_controller.hpp>
-#include <eos/chain/producer_object.hpp>
-#include <eos/chain/exceptions.hpp>
+#include <eosio/chain/chain_controller.hpp>
```
Comment (wanderingbort) [inline tests/common/database_fixture.hpp]: Is this file in use for anything?  can it be deleted.

File: tests/common/macro_support.hpp [hunk 16]
```diff
@@ -10,7 +10,7 @@
  * @file Contains support macros for the testcase helper macros. These macros are implementation details, and thus
  * should not be used directly. Use their frontends instead.
  */
-
+#if 0
```
Comment (wanderingbort) [inline tests/common/macro_support.hpp]: same here, can we just delete this file now?   nothing shoudl be using the old Macros anymore

File: tests/wasm_tests/wasm_tests.cpp [hunk 17]
```diff
@@ -226,44 +202,105 @@ BOOST_FIXTURE_TEST_CASE( abi_from_variant, tester ) try {
 
 } FC_LOG_AND_RETHROW() /// prove_mem_reset
 
-BOOST_FIXTURE_TEST_CASE( test_api_bootstrap, tester ) try {
+BOOST_FIXTURE_TEST_CASE( test_deferred_failure, tester ) try {
```
Comment (wanderingbort) [inline tests/wasm_tests/wasm_tests.cpp]: pretty sure this test was moved to the currency tests as it required currency... is this file disabled from running?

File: contracts/eosiolib/checktime.h (hunk 5 above)
Comment (spoonincode) [reply contracts/eosiolib/checktime.h]: Yeah I don't think we should expose it (we have a ticket somewhere where we intend to actively refuse linkage to it in the future). This call is implementation detail.

Comment (larryk85) [reply contracts/eosiolib/checktime.h]: Yeah, this was an oversight on my part.  I exposed it so that I could see what it was actually doing and forgot to delete it from the header.

File: contracts/test_api/test_privileged.cpp (hunk 7 above)
Comment (larryk85) [reply contracts/test_api/test_privileged.cpp]: These tests are not finished yet, and I didn't mean to include these with this PR.
